test/unit/org/apache/cassandra/utils/SortingIteratorTest.java

```diff
@@ -0,0 +1,417 @@
+    public void testSortingIterator_withRandomData() {
+        Random random = new Random();
+        int size = 10000;
+        List<Integer> data = new ArrayList<>();
+        for (int i = 0; i < size; i++) {
+            data.add(random.nextInt(size));
+        }
+
+        Iterator<Integer> iterator = makeIterator(Comparator.naturalOrder(), data);
+        List<Integer> sorted = new ArrayList<>();
+        while (iterator.hasNext()) {
+            sorted.add(iterator.next());
+        }
+
+        List<Integer> expected = new ArrayList<>(data);
+        Collections.sort(expected);
+
+        assertEquals(expected, sorted);
+    }
+
+    @Test
+    public void testSortingIterator_skipToRandomData() {
+        Random random = new Random();
+        int size = 10000;
+        List<Integer> data = new ArrayList<>();
+        for (int i = 0; i < size; i++) {
+            data.add(random.nextInt(size));
+        }
+
+        List<Integer> sorted = new ArrayList<>(data);
+        Collections.sort(sorted);
+        List<Integer> expected = new ArrayList<>();
+
+        var iterator = makeIterator(Comparator.naturalOrder(), data);
+        List<Integer> iterated = new ArrayList<>();
+
+        int skipDistanceMax = 200;
+        for (int i = 0; i < size; i += random.nextInt(skipDistanceMax))
+        {
+            int targetMax = sorted.get(i);
+            int targetMin = i > 0 ? sorted.get(i - 1) : 0;
+            if (targetMin == targetMax)
+                continue;
+            int target = random.nextInt(targetMax - targetMin) + targetMin + 1; // (targetMin + 1; targetMax]
+            iterator.skipTo(target);
+            for (int c = random.nextInt(5); c >= 0; --c)
+            {
+                if (i >= size)
+                {
+                    assert !iterator.hasNext();
+                    break;
+                }
+                iterated.add(iterator.next());
+                expected.add(sorted.get(i++));
+            }
+        }
+
+        assertEquals(expected, iterated);
+    }
+
+    @Test
+    public void testIteratorRemovesDuplicates() {
+        List<Integer> data = Arrays.asList(1, 2, 2, 3, 4, 4, 4, 5);
+        SortingIterator<Integer> iterator = SortingIterator.createDeduplicating(Comparator.naturalOrder(), data);
+
+        List<Integer> result = new ArrayList<>();
+        while (iterator.hasNext()) {
+            result.add(iterator.next());
+        }
+
+        assertEquals(Arrays.asList(1, 2, 3, 4, 5), result);
+    }
+
+    @Test
+    public void testSkipTo_withDuplicates() {
+        List<Integer> data = Arrays.asList(1, 2, 2, 3, 4, 4, 4, 5);
+        SortingIterator<Integer> iterator = SortingIterator.createDeduplicating(Comparator.naturalOrder(), data);
+
+        iterator.skipTo(3);
+        assertTrue(iterator.hasNext());
+        assertEquals((Integer) 3, iterator.next());
+
+        iterator.skipTo(4);
+        assertTrue(iterator.hasNext());
+        assertEquals((Integer) 4, iterator.next());
```

Comment:
suggestion: Should check if it skipped to the first 4. So check if we get the 4 three times.

Reply:
Fixed